txnkv/transaction/prewrite.go

```diff
@@ -226,286 +227,368 @@ func (action actionPrewrite) handleSingleBatch(
+				zap.Stringer("lock", lock),
+			)
+			logged[lock.TxnID] = struct{}{}
 		}
-		resolveLockOpts := txnlock.ResolveLocksOptions{
-			CallerStartTS:            c.startTS,
-			Locks:                    locks,
-			Detail:                   &c.getDetail().ResolveLock,
-			PessimisticRegionResolve: true,
+		// If an optimistic transaction encounters a lock with larger TS, this transaction will certainly
+		// fail due to a WriteConflict error. So we can construct and return an error here early.
+		// Pessimistic transactions don't need such an optimization. If this key needs a pessimistic lock,
+		// TiKV will return a PessimisticLockNotFound error directly if it encounters a different lock. Otherwise,
+		// TiKV returns lock.TTL = 0, and we still need to resolve the lock.
+		if lock.TxnID > handler.committer.startTS && !handler.committer.isPessimistic {
+			return nil, tikverr.NewErrWriteConflictWithArgs(
+				handler.committer.startTS,
+				lock.TxnID,
+				0,
+				lock.Key,
+				kvrpcpb.WriteConflict_Optimistic,
+			)
 		}
-		resolveLockRes, err := c.store.GetLockResolver().ResolveLocksWithOpts(bo, resolveLockOpts)
+		locks = append(locks, lock)
+	}
+	return locks, nil
+}
+
+// resolveLocks resolves locks.
+func (handler *prewrite1BatchReqHandler) resolveLocks(locks []*txnlock.Lock) error {
+	if handler.resolvingRecordToken == nil {
+		token := handler.committer.store.GetLockResolver().RecordResolvingLocks(locks, handler.committer.startTS)
+		handler.resolvingRecordToken = &token
+	} else {
+		handler.committer.store.GetLockResolver().UpdateResolvingLocks(locks, handler.committer.startTS, *handler.resolvingRecordToken)
+	}
+
+	resolveLockOpts := txnlock.ResolveLocksOptions{
+		CallerStartTS:            handler.committer.startTS,
+		Locks:                    locks,
+		Detail:                   &handler.committer.getDetail().ResolveLock,
+		PessimisticRegionResolve: true,
+	}
+	resolveLockRes, err := handler.committer.store.GetLockResolver().ResolveLocksWithOpts(handler.bo, resolveLockOpts)
+	if err != nil {
+		return err
+	}
+	msBeforeExpired := resolveLockRes.TTL
+	if msBeforeExpired > 0 {
+		err = handler.bo.BackoffWithCfgAndMaxSleep(
+			retry.BoTxnLock,
+			int(msBeforeExpired),
+			errors.Errorf("2PC prewrite lockedKeys: %d", len(locks)),
+		)
 		if err != nil {
 			return err
 		}
-		msBeforeExpired := resolveLockRes.TTL
-		if msBeforeExpired > 0 {
-			err = bo.BackoffWithCfgAndMaxSleep(
-				retry.BoTxnLock,
-				int(msBeforeExpired),
-				errors.Errorf("2PC prewrite lockedKeys: %d", len(locks)),
+	}
+	return nil
+}
+
+// handleSingleBatchSucceed handles the response when the prewrite request is successful.
+func (handler *prewrite1BatchReqHandler) handleSingleBatchSucceed(prewriteResp *kvrpcpb.PrewriteResponse) error {
+	// Clear the RPC Error since the request is evaluated successfully.
+	handler.sender.SetRPCError(nil)
+
+	// Update CommitDetails
+	reqDuration := time.Since(handler.begin)
```

Comment:
`handler.begin` is not [`reqBegin`](https://github.com/tikv/client-go/blob/6e324f2e07c93680aa08a7ca0a8dca89a3f70e30/txnkv/transaction/prewrite.go#L350).